File: MessageHandler/MessageHandler.py
import threading
from common import Config, getTime, printerror, printsuccess, printdebug
from confluent_kafka import Consumer, Producer
from json import loads, dumps
from datetime import datetime

class MessageHandler:
    def __init__(self):
        pass
    
    def run(self):
        def __get_header(headers_tuple, key):
            for header in headers_tuple:
                if header[0] == key:
                    return header[1]
            return None

        def __set_header(headers_tuples, new_tuples):
            new_keys = [i[0] for i in new_tuples]
            return new_tuples + [i for i in headers_tuples if i[0] not in new_keys]

        def __get_topic(headers_tuple):
            try:
                headers = dict((k, v.decode('utf-8')) for k, v in headers_tuple)
                if (not headers) or (not headers['topic']):
                    return False
                
                if (not headers['time']):
                    return headers['topic']

                time_diff = (datetime.strptime(headers['time'], Config.get('sm_time_format')) - datetime.now()).total_seconds()
                
                # If the message is scheduled for a time too small to bucket, we send it to the provided topic
                if time_diff < Config.get('sm_miniumum_delay'):
                    return headers['topic']
                
                # If the message is scheduled for the future (+ bucket process time), then we need to bucket it in the associated bucket
                for bucket in reversed(Config.get('bucket_object_list')):
                    if time_diff > (1 + Config.get('sm_bucket_process_time_fraction')) * bucket['lower']:
                        return bucket['name']

                return headers['topic']
            except Exception as e:
                printerror(f'Unable to get topic name from headers: {headers}')
                printerror(e)
                return False

        while True:
            try:
                consumer = Consumer({
                    'bootstrap.servers':    Config.get('kafka_server'),
                    'group.id':             Config.get('sm_mh_consumer_group_id'),
                    'enable.auto.commit':   True,
                    'auto.offset.reset':    'earliest',
                })
                consumer.subscribe([Config.get('sm_topic')])
                
                producer = Producer({
                    'bootstrap.servers':    Config.get('kafka_server')})
                
                message_count = 0
                while True:
                    message = consumer.poll(timeout=Config.get('worker_consumer_timeout'))
                    
                    if message is None or message.error():
                        message and message.error() and printerror(f'{message.error()}')
                        producer.flush()
                    
                    else:
                        topic = __get_topic(message.headers())

                        if topic:
                            message_count += 1
                            message.set_headers(__set_header(message.headers(), [(Config.get('sm_header_mh_timestamp_key'), bytes(getTime(), 'utf-8')),]))

                            producer.produce(
                                topic   = topic,
                                value   = message.value(),
                                key     = message.key(),
                                headers = message.headers(),
                            )

                            (message_count % 100 == 0) and producer.flush()
                    

            except Exception as e:
                try:
                    printerror(f'MessageHandler.run(): Trying to close consumer and producer')
                    producer.flush()
                    consumer.close()
                except Exception as e:
                    printerror(f'Unable to close consumer and producer: {e}')
                printerror(e)
                printerror(f'Worker PID: {threading.get_native_id()} terminated. Restarting...')

if __name__ == '__main__':

    mh = MessageHandler()
    mh.run()

File: MessageHandler/src/messagehandler.rs
use rdkafka::config::ClientConfig;
use rdkafka::consumer::base_consumer::BaseConsumer;
use rdkafka::message::{Headers, BorrowedHeaders};
use rdkafka::producer::{FutureProducer, FutureRecord};
use futures::executor;

use rdkafka::consumer::Consumer;
use rdkafka::message::Message;

use serde_json::Value as Json;

use crate ::common::*;

pub struct MessageHandler {
    config: Json,
}

impl MessageHandler {
    pub fn new(config: &Json) -> MessageHandler {
        MessageHandler {
            config: config.clone(),
        }
    }
    pub fn run(&self) {
        let __run = | config : &Json | -> Result<(), String> { 
            let __get_topic_name = |config: &Json, headers: &BorrowedHeaders| -> Option<String> {
                let mut sm_topic_name = None;
                let mut sm_time = None;
        
                for i in 0..headers.count() {
                    let header = headers.get(i).unwrap_or_default();
                    if header.0 == "topic" {
                        sm_topic_name = Some(u8_to_str(header.1));
                    } else if header.0 == "time" {
                        sm_time = Some(u8_to_str(header.1));
                    }
                }
                if sm_topic_name == None {
                    return None;
                }
                if sm_time == None {
                    return sm_topic_name;
                }
        
                let sm_time = sm_time.unwrap();
                let sm_time = chrono::NaiveDateTime::parse_from_str(sm_time.as_str(), get_string(&config, "sm_time_format").as_str()).unwrap();
                let time_diff = (sm_time.time() - chrono::Local::now().time()).num_seconds();
                
                if time_diff < -1 * get_number(&config, "sm_miniumum_delay") as i64 {
                    let error_msg = format!("MessageHandler DELAY: {} secs : {:?}", time_diff, sm_time);
                    print_warning(error_msg.as_str());
                    return sm_topic_name;
                }
                if time_diff < get_number(&config, "sm_miniumum_delay") as i64 {
                    return sm_topic_name;
                }
                let bucket_list = get_list(&config, "bucket_object_list").into_iter().rev().collect::<Vec<_>>();
                for bucket in bucket_list.iter() {
                    let b_time = (1.0 + get_float(&config, "sm_bucket_process_time_fraction")) * get_float(&bucket, "lower");
                    if time_diff > b_time as i64 {
                        return Some(get_string(bucket, "name"));
                    }
                }
                sm_topic_name
            };

            let __get_header = |headers: &BorrowedHeaders, key: &str| -> Option<String> {
                for i in 0..headers.count() {
                    let header = headers.get(i).unwrap_or_default();
                    if header.0 == key {
                        return Some(u8_to_str(header.1));
                    }
                }
                None
            };

            let consumer : BaseConsumer = ClientConfig::new()
                    .set("group.id", get_string(&config, "sm_mh_consumer_group_id"))
                    .set("bootstrap.servers", get_string(&config, "kafka_server"))
                    .set("enable.auto.commit", "true")
                    .set("auto.offset.reset", "earliest").create().expect("MessageHandler.__run(): Consumer creation error");

            let producer : &FutureProducer = &ClientConfig::new()
                .set("bootstrap.servers", get_string(&config, "kafka_server"))
                .create().expect("MessageHandler.__run(): Producer creation error");

            consumer.subscribe(&[get_string(&config, "sm_topic").as_str()]).expect("MessageHandler.__run(): Subscribe error");
            
            loop {
                match consumer.poll(std::time::Duration::from_secs(get_number(&config, "worker_consumer_timeout"))) {
                    Some(message) => {
                        let message = message.expect("MessageHandler.__run(): Message Parse Error");
                        
                        let headers = message.headers().unwrap();

                        let topic_name = __get_topic_name(&config, &headers);
                        if topic_name == None {
                            continue;
                        }
                        let topic_name = topic_name.unwrap();
                        
                        let msg_headers = headers.detach().add(&get_string(&config, "sm_header_mh_timestamp_key"), get_time().as_bytes());

                        match executor::block_on(producer.send(
                            FutureRecord::to(&topic_name)
                                .payload(message.payload().unwrap_or_default())
                                .key(message.key().unwrap_or_default())
                                .headers(msg_headers),
                            std::time::Duration::from_secs(0),
                        )) {
                            Ok(_) => {},
                            Err(e) => {
                                let err_msg = format!("MessageHandler.__run(): Producer send error: {:?}", e);
                                print_error(&err_msg);
                                return Err(err_msg);
                            }
                        }
                    },
                    None => {
                        continue;
                    }
                }
            }
        };

        loop {
            match __run(&self.config) {
                Ok(_) => {},
                Err(e) => {
                    print_error(&format!("MessageHandler::run(): {}", e));
                    print_error("MessageHandler::run() failed. Trying to restart the process now...");
                }
            }
            std::thread::sleep(std::time::Duration::from_secs(1));
        }
    }
}
